### ISSUE TITLE
Allow `IntegratedGradients.target` to be inferred for classification models with class score

### ISSUE THREAD

**Issue by sakoush** [SeldonIO/alibi#492]
As part of ongoing work to integrate `alibi` runtime in `mlserver`, we realised that for `IntegratedGradients` if the inference model outputs a class score (e.g mnist model handwritten digit recognition), a target index is required to be passed to `IntegratedGradients.explain` 

i.e.

`target = model(X_test_sample).numpy().argmax(axis=1)`

I see why it is required but at the same time I am not sure how we can easily handle this kind of custom code in production / deployment. 

`seldon-core` v1 does have it fixed in the IG explainer model wrapper but it is restrictive i.e. we are not sure it will generalises to all use cases.

Having a discussion with @jklaise  we could consider some changes to alibi to make it easier in this case.

**Comment by jklaise** [comment]
Having thought more about it, it seems to me the cleanest and most Pythonic way to solve this is via callbacks. Instead of extending the definition of `target` to take a restricted set of strings such as `argmax` that map to internal functions, allow the user passing the function in question directly (and maybe disambiguate the semantics by introducing another kwarg `target_fn` with the expectation that only one of `target` or `target_fn` can be not `None`). In this example `target_fn=np.argmax` (module some `partial`s with axis), but the only requirement on `target_fn` is that it maps (batched) vectors to scalars.

 (There is a lot of precedent for callbacks as it leads to powerful Pythonic code, but in particular, [captum](https://captum.ai/api/neuron.html) use this for their `Neuron*` explainers in exactly the same way - user defined callable that selects the neuron of interest within the layer.

Of course, this brings us to the old issue of how to specify callbacks from outside of Python, e.g. in a config file. It would be silly for someone to pickle a built-in `np.argmax` function just so that together with a config file the explainer can be fully spec'ed. This leads me to think more that we should embrace [catalogue](https://github.com/explosion/catalogue) internally, create internal `alibi` registries for functions that can be used as callbacks in `alibi` code and then enable users to refer to them via their registered strings either in config files (which resolve to the `alibi` function at runtime) or accepting the registered strings to the explainer constructor.

Would be keen to hear your thoughts @ascillitoe @sakoush @arnaudvl.

**Comment by ascillitoe** [comment]
I agree, `target_fn` seems like a really nice option. The similarity with `preprocess_fn` in `alibi-detect` means we can of course reuse lots of the functionality between the two libraries, but more importantly I'd hope the similarity across the libraries would improve the user experience here.

**Comment by sakoush** [comment]
> I think the overarching theme is that we should not shy away from callbacks in public Python APIs but embrace them (alibi-detect already does this extensively), at the same time finding a solution for how to specify them via string keys from outside of Python.
yes this is great. This pattern can be used in other areas where we need it such as in `AnchorText` where the user can specify a string for  `language_model` that are registered from within alibi.

**Comment by ascillitoe** [comment]
Brief not re using a `catalogue` type function registry for `target_fn`, we will need to think about whether we want the explainers themselves to accept registry strings and resolve to a callable (in which case the explainer signatures will change). Or do we adopt the current Alibi Detect strategy where registered functions e.g. `preprocess_fn` can only be specified in a config file, and separate utility functions first resolve the string registry before passing the callable to the detector/explainer (in which case the detector/explainer signature stays the same). 

@jklaise is #523 going to close this issue? If so shall we continue this discussion elsewhere?

**Comment by jklaise** [comment]
I like the idea of separating the public API so that strings would be only accepted via config and implementation work on callables. This is a bit fuzzy though as we already use string keys to refer to functions elsewhere, e.g. `segmentation_fn` in `AnchorImage`.

We can keep this issue open but likely carry on the discussion relevant to string keys in a new one once we get to it.
